tests/agent_server/deploy/e2e/test_scheduler_initialization.py

```diff
@@ -0,0 +1,135 @@
+            "id": f"test::Resource[agent1,key=key3],v={version}",
+            "requires": [f"test::Resource[agent1,key=key2],v={version}"],
+            "purged": False,
+            "send_event": False,
+        },
+    ]
+    # Deploy and release version
+    await clienthelper.put_version_simple(version=version, resources=resources)
+    result = await client.release_version(environment, version)
+    assert result.code == 200
+    await clienthelper.wait_for_deployed()
+
+    for rid, expected_status in [
+        ("test::Resource[agent1,key=key1]", const.ResourceState.deployed),
+        ("test::Resource[agent1,key=key2]", const.ResourceState.failed),
+        ("test::Resource[agent1,key=key3]", const.ResourceState.skipped),
+    ]:
+        result = await client.resource_details(tid=environment, rid=rid)
+        assert result.code == 200
+        assert result.result["data"]["status"] == expected_status.value
+
+    result = await client.get_resource_actions(
+        tid=environment,
+        resource_type="test::Resource",
+        agent="agent1",
+        exclude_changes=[const.Change.nochange.value],
+    )
+    assert result.code == 200
+    assert len(result.result["data"]) == 1
+    assert result.result["data"][0]["resource_version_ids"] == ["test::Resource[agent1,key=key1],v=1"]
+
+    result = await client.get_resource_actions(
+        tid=environment,
+        resource_type="test::Resource",
+        agent="agent1",
+    )
+    assert result.code == 200
+
+    # Pause the agent to stop the scheduler
+    result = await client.agent_action(tid=environment, name=const.AGENT_SCHEDULER_ID, action=const.AgentAction.pause.value)
+    assert result.code == 200
+    result = await client.agent_action(tid=environment, name=const.AGENT_SCHEDULER_ID, action=const.AgentAction.unpause.value)
+    assert result.code == 200
+    await utils.retry_limited(utils.is_agent_done, scheduler=agent.scheduler, agent_name="agent1", timeout=10, interval=0.05)
+
+    for rid, expected_status in [
+        ("test::Resource[agent1,key=key1]", const.ResourceState.deployed),
+        ("test::Resource[agent1,key=key2]", const.ResourceState.deployed),
+        ("test::Resource[agent1,key=key3]", const.ResourceState.deployed),
+    ]:
+        result = await client.resource_details(tid=environment, rid=rid)
+        assert result.code == 200
+        assert result.result["data"]["status"] == expected_status.value, f"{rid} has unexpected state"
+
+    result = await client.get_resource_actions(
+        tid=environment,
+        resource_type="test::Resource",
+        agent="agent1",
+        exclude_changes=[const.Change.nochange.value],
+    )
+    assert result.code == 200
+
+    result = await client.get_resource_actions(
+        tid=environment,
+        resource_type="test::Resource",
+        agent="agent1",
+        exclude_changes=[const.Change.nochange.value],
+    )
+    assert result.code == 200
+    assert len(result.result["data"]) == 3
```

Comment:
This is unexpected? I would expect 5?

1. one for key1, first deploy
2. one for key2, first deploy
3. one for key3, first deploy
4. one for key2, second deploy
5. one for key3, second deploy

Reply:
The call to `get_resource_actions()` filters out the "nochange" actions. Failure events are logged as "nochange". As such the following happens:

First deploy (before restart):

* Only `test::Resource[agent1,key=key1]` gets deployed successfully so we only have one resource action

Second deploy (after restart):

* `test::Resource[agent1,key=key1]` was already successfully deployed, so the recovery process of the scheduler doesn't schedule it again.
* `test::Resource[agent1,key=key2]` and ``test::Resource[agent1,key=key3]`` are deployed successfully.

So after the restart there are three resource actions without the "nochange" type. `test::Resource[agent1,key=key1]` (from before the restart) and `test::Resource[agent1,key=key2]` and `test::Resource[agent1,key=key3]` (after the restart). Do I need to document this better. Or do I need to remove the filter on nochange?